ms-windows/build-tools.php

```diff
@@ -0,0 +1,1561 @@
+			if (in_array("all", $copymodes) || in_array("test", $copymodes))  CopyBuildFiles($indir, $outdir2, "test", true);
+			if (in_array("all", $copymodes) || in_array("tools", $copymodes))  CopyBuildFiles($indir, $outdir2, "tools", true);
+
+			CopyBuildFiles($indir, $outdir2, "", false, '/(\.exp|\.lib|\.def|\.dll|\.pdb)$/');
+
+
+			// Build comleted.
+			$ts2 = time();
+			$secs = $ts2 - $ts;
+			$mins = (int)($secs / 60);
+			$secs %= 60;
+
+			echo $echoprefix . " Finished build at " . date("Y-m-d H:i:s") . ".  Build time:  " . $mins . " min " . $secs . " sec\n";
+		}
+
+		// Run first build.  This build takes much longer due to also building the test suite.
+		if (!is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/dll_MD") || !is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/inc"))
+		{
+			RunBuild($ver, $arch, $baseenv, 1, $pathmap, $archinfo, "shared", "/MD", "dll_MD", array("all"));
+
+			// Copy the include files.
+			@mkdir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/inc/openssl", 0775, true);
+			$dir = opendir($pathmap["[[SOURCE_TEMP_DIR]]"] . "/include/openssl");
+			if ($dir)
+			{
+				while (($file = readdir($dir)) !== false)
+				{
+					if (strtolower(substr($file, -2)) === ".h")
+					{
+						// Convert line endings to DOS line endings.
+						$data = file_get_contents($pathmap["[[SOURCE_TEMP_DIR]]"] . "/include/openssl/" . $file);
+
+						$data = str_replace("\r\n", "\n", $data);
+						$data = str_replace("\r", "\n", $data);
+						$data = str_replace("\n", "\r\n", $data);
+
+						file_put_contents($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/inc/openssl/" . $file, $data);
+					}
+				}
+
+				closedir($dir);
+			}
+
+			// Copy ms/applink.c too.
+			// Convert line endings to DOS line endings.
+			$data = file_get_contents($pathmap["[[SOURCE_TEMP_DIR]]"] . "/ms/applink.c");
+
+			$data = str_replace("\r\n", "\n", $data);
+			$data = str_replace("\r", "\n", $data);
+			$data = str_replace("\n", "\r\n", $data);
+
+			file_put_contents($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/inc/openssl/applink.c", $data);
+		}
+
+		// Run the remaining builds.
+		if (!is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/dll_MT"))  RunBuild($ver, $arch, $baseenv, 2, $pathmap, $archinfo, "shared no-tests", "/MT", "dll_MT", array("apps"));
+		if (!is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/dll_MDd"))  RunBuild($ver, $arch, $baseenv, 3, $pathmap, $archinfo, "--debug shared no-tests", "/MDd", "dll_MDd", array());
+		if (!is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/dll_MTd"))  RunBuild($ver, $arch, $baseenv, 4, $pathmap, $archinfo, "--debug shared no-tests", "/MTd", "dll_MTd", array());
+
+		if (!is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/static_MD"))  RunBuild($ver, $arch, $baseenv, 5, $pathmap, $archinfo, "no-shared no-tests", "/MD", "static_MD", array("apps"));
+		if (!is_dir($pathmap["[[BUILD_OUTPUT_DIR]]"] . "/static_MT"))  RunBuild($ver, $arch, $baseenv, 6, $pathmap, $archinfo, "no-shared no-tests", "/MT", "static_MT", array("apps"));
```

Comment:
The primary purpose of builds 2-4 are to get linkable .lib files, not the DLLs that also get generated.  I've always encouraged devs to use BIOs over FILEs to avoid VC++ runtime conflicts and also improve cross-platform compatibility.  Those generated .libs will load the /MD compiled DLLs but the VS linker is very picky about how .lib files are linked despite the Windows loader not caring.  Is it playing with fire?  A little.  But I've never heard any complaints once the BIO APIs get used - usually "thanks" happens instead.

The primary purpose of builds 5 and 6 are to get a statically linked `openssl.exe` depending on how "static" you, or someone else using the build tools, wants.  A fully runtime-linked `openssl.exe` `/MT` build might crash for some operations.  I tried to do that a number of years ago to try to eliminate the VC++ runtime dependency and ended up very rapidly going back to `/MD` when several people came screaming out of the woodwork about crashes.  I couldn't revert fast enough.

The `no-tests` skips building the test suite altogether because building the test suite adds several minutes to each build.  When you are doing 48 separate builds, every second multiplies the time taken.  I'm not currently _running_ the test suite as part of the binary build process as I assume CI/CD is doing that already.  Generating them as binaries is a matter of convenience for end users who want to run the tests for themselves on their systems.

I suppose the real issue here is that there are 6 builds being done per architecture per version for specific reasons.  It might make more sense to go back and evaluate the generated makefile to see if it can produce everything required in just one build cycle by executing the linker step several additional times.  That would reduce 48 builds down to just 8 builds and take a fraction of the time it currently takes at the small cost of being just slightly longer per build.  I've always worked within the limitations presented by the `Configure` script.  We can always work toward optimizing the build process.